File: web/src/api/endpoints/embeddingConfig.ts
/**
 * Embedding Configuration API endpoints
 */

import apiClient from '../client';
import { API_ENDPOINTS } from '../../utils/constants';
import type {
  EmbeddingConfigPublic,
  EmbeddingConfigsPublic,
  EmbeddingConfigCreate,
  EmbeddingConfigUpdate,
  EmbeddingConfigTest,
  EmbeddingModelsFetchRequest,
  EmbeddingConfigTestResult,
  EmbeddingProvidersInfo,
  AvailableEmbeddingModels,
} from '../types/embeddingConfig.types';
import type { Message } from '../types/common.types';

export const embeddingConfigApi = {
  /**
   * Get all Embedding configurations for the current user
   */
  list: async (): Promise<EmbeddingConfigsPublic> => {
    const response = await apiClient.get<EmbeddingConfigsPublic>(
      API_ENDPOINTS.EMBEDDING_CONFIG.LIST
    );
    return response.data;
  },

  /**
   * Get a single Embedding configuration by ID
   */
  get: async (id: string): Promise<EmbeddingConfigPublic> => {
    const response = await apiClient.get<EmbeddingConfigPublic>(
      API_ENDPOINTS.EMBEDDING_CONFIG.GET(id)
    );
    return response.data;
  },

  /**
   * Create a new Embedding configuration
   */
  create: async (
    data: Omit<EmbeddingConfigCreate, 'user_id'>
  ): Promise<EmbeddingConfigPublic> => {
    const response = await apiClient.post<EmbeddingConfigPublic>(
      API_ENDPOINTS.EMBEDDING_CONFIG.CREATE,
      data
    );
    return response.data;
  },

  /**
   * Update an existing Embedding configuration
   */
  update: async (
    id: string,
    data: EmbeddingConfigUpdate
  ): Promise<EmbeddingConfigPublic> => {
    const response = await apiClient.patch<EmbeddingConfigPublic>(
      API_ENDPOINTS.EMBEDDING_CONFIG.UPDATE(id),
      data
    );
    return response.data;
  },

  /**
   * Delete an Embedding configuration
   */
  delete: async (id: string): Promise<Message> => {
    const response = await apiClient.delete<Message>(
      API_ENDPOINTS.EMBEDDING_CONFIG.DELETE(id)
    );
    return response.data;
  },

  /**
   * Test an Embedding configuration
   */
  test: async (data: EmbeddingConfigTest): Promise<EmbeddingConfigTestResult> => {
    const response = await apiClient.post<EmbeddingConfigTestResult>(
      API_ENDPOINTS.EMBEDDING_CONFIG.TEST,
      data
    );
    return response.data;
  },

  /**
   * Fetch available embedding models from the current provider configuration
   */
  fetchModels: async (data: EmbeddingModelsFetchRequest): Promise<AvailableEmbeddingModels> => {
    const response = await apiClient.post<AvailableEmbeddingModels>(
      API_ENDPOINTS.EMBEDDING_CONFIG.FETCH_MODELS,
      data
    );
    return response.data;
  },

  /**
   * Set an Embedding configuration as default
   */
  setDefault: async (id: string): Promise<EmbeddingConfigPublic> => {
    const response = await apiClient.post<EmbeddingConfigPublic>(
      API_ENDPOINTS.EMBEDDING_CONFIG.SET_DEFAULT(id)
    );
    return response.data;
  },

  /**
   * Get all available Embedding providers
   */
  getProviders: async (): Promise<EmbeddingProvidersInfo> => {
    const response = await apiClient.get<EmbeddingProvidersInfo>(
      API_ENDPOINTS.EMBEDDING_CONFIG.PROVIDERS
    );
    return response.data;
  },
};

File: slidegen/api/routers/embedding_config.py
import uuid
from typing import Any

from fastapi import APIRouter, HTTPException
from loguru import logger
from sqlmodel import col, func, select

from slidegen.api.deps import CurrentUser, SessionDep
from slidegen.exceptions import NotFoundError, ParamsCheckError
from slidegen.models.embedding_config import (
    DEFAULT_EMBEDDING_CONFIGS,
    EmbeddingConfigCreate,
    EmbeddingConfigModel,
    EmbeddingConfigPublic,
    EmbeddingConfigsPublic,
    EmbeddingConfigUpdate,
    EmbeddingProvider,
)
from slidegen.models.user import Message
from slidegen.schemas.embedding_config import (
    EMBEDDING_PROVIDER_INFO,
    AvailableEmbeddingModels,
    EmbeddingConfigTest,
    EmbeddingConfigTestResult,
    EmbeddingProvidersInfo,
)
from slidegen.services.factories.embedding_factory import EmbeddingFactory

router = APIRouter()


def _is_masked_secret(value: str | None) -> bool:
    if value is None:
        return False
    return bool(value) and value.startswith("***")


@router.get("/providers", response_model=EmbeddingProvidersInfo)
async def get_providers() -> EmbeddingProvidersInfo:
    """Get all supported embedding providers information"""
    return EmbeddingProvidersInfo(providers=list(EMBEDDING_PROVIDER_INFO.values()))


@router.get("/providers/{provider}/models", response_model=AvailableEmbeddingModels)
async def get_provider_models(provider: EmbeddingProvider) -> AvailableEmbeddingModels:
    """Get available models for a specific provider"""
    if provider not in DEFAULT_EMBEDDING_CONFIGS:
        raise HTTPException(status_code=404, detail=f"Provider {provider} does not have preset model configurations")

    models = DEFAULT_EMBEDDING_CONFIGS[provider]
    return AvailableEmbeddingModels(provider=provider, models=models)


@router.post("/test", response_model=EmbeddingConfigTestResult)
async def test_embedding_config(
    config: EmbeddingConfigTest,
    session: SessionDep,
    current_user: CurrentUser,
) -> EmbeddingConfigTestResult:
    """Test if the embedding configuration is valid"""
    if config.config_id and (not config.api_key or _is_masked_secret(config.api_key)):
        stored_config = await session.get(EmbeddingConfigModel, config.config_id)
        if not stored_config or stored_config.user_id != current_user.id:
            raise NotFoundError("Configuration not found")

        config = config.model_copy(update={"api_key": stored_config.api_key})

    # First validate configuration basic parameters
    is_valid, error_msg = EmbeddingFactory.validate_config(config)
    if not is_valid:
        return EmbeddingConfigTestResult(success=False, error=error_msg)

    # Perform actual test
    result = await EmbeddingFactory.test_embedding_config(config)
    return result


@router.get("/", response_model=EmbeddingConfigsPublic)
async def get_embedding_configs(
    session: SessionDep, current_user: CurrentUser, skip: int = 0, limit: int = 100
) -> EmbeddingConfigsPublic:
    """Get list of embedding configurations for the current user"""
    count_statement = (
        select(func.count()).select_from(EmbeddingConfigModel).where(EmbeddingConfigModel.user_id == current_user.id)
    )
    count = (await session.execute(count_statement)).scalar_one()

    statement = (
        select(EmbeddingConfigModel)
        .where(EmbeddingConfigModel.user_id == current_user.id)
        .offset(skip)
        .limit(limit)
        .order_by(col(EmbeddingConfigModel.is_default).desc(), col(EmbeddingConfigModel.create_time).desc())
    )
    configs = (await session.execute(statement)).scalars().all()

    return EmbeddingConfigsPublic(data=configs, count=count)


@router.post("/", response_model=EmbeddingConfigPublic)
async def create_embedding_config(
    *, session: SessionDep, current_user: CurrentUser, config_in: EmbeddingConfigCreate
) -> Any:
    """Create new embedding configuration"""
    # Validate configuration
    is_valid, error_msg = EmbeddingFactory.validate_config(config_in)
    if not is_valid:
        logger.warning(f"Invalid embedding config for user {current_user.id}: {error_msg}")
        raise ParamsCheckError(error_msg)

    # If set as default configuration, cancel other default configurations
    if config_in.is_default:
        # Cancel other default configurations for the current user
        statement = select(EmbeddingConfigModel).where(
            EmbeddingConfigModel.user_id == current_user.id, EmbeddingConfigModel.is_default
        )
        existing_defaults = (await session.execute(statement)).scalars().all()
        for default_config in existing_defaults:
            default_config.is_default = False
            session.add(default_config)

    # Create configuration
    config = EmbeddingConfigModel.model_validate(
        {
            **config_in.model_dump(),
            "user_id": current_user.id,
        }
    )
    session.add(config)
    await session.commit()
    await session.refresh(config)

    logger.info(
        f"Created embedding config {config.id} for user {current_user.id}: {config.provider} - {config.model_id}"
    )

    return config


@router.get("/{config_id}", response_model=EmbeddingConfigPublic)
async def get_embedding_config(session: SessionDep, current_user: CurrentUser, config_id: uuid.UUID) -> Any:
    """Get specified embedding configuration"""
    config = await session.get(EmbeddingConfigModel, config_id)
    if not config:
        raise NotFoundError("Configuration not found")

    if config.user_id != current_user.id:
        raise NotFoundError("Configuration not found")

    return config


@router.patch("/{config_id}", response_model=EmbeddingConfigPublic)
async def update_embedding_config(
    *,
    session: SessionDep,
    current_user: CurrentUser,
    config_id: uuid.UUID,
    config_in: EmbeddingConfigUpdate,
) -> Any:
    """Update embedding configuration"""
    config = await session.get(EmbeddingConfigModel, config_id)
    if not config:
        raise NotFoundError("Configuration not found")

    if config.user_id != current_user.id:
        raise NotFoundError("Configuration not found")

    # If set as default configuration, cancel other default configurations
    if config_in.is_default:
        statement = select(EmbeddingConfigModel).where(
            EmbeddingConfigModel.user_id == current_user.id,
            EmbeddingConfigModel.is_default,
            EmbeddingConfigModel.id != config_id,
        )
        existing_defaults = (await session.execute(statement)).scalars().all()
        for default_config in existing_defaults:
            default_config.is_default = False
            session.add(default_config)

    # Update configuration
    config_data = config_in.model_dump(exclude_unset=True)
    if "api_key" in config_data and _is_masked_secret(config_data["api_key"]):
        config_data.pop("api_key")
    config.sqlmodel_update(config_data)

    # Validate updated configuration
    is_valid, error_msg = EmbeddingFactory.validate_config(config)
    if not is_valid:
        raise ParamsCheckError(error_msg)

    session.add(config)
    await session.commit()
    await session.refresh(config)

    return config


@router.delete("/{config_id}", response_model=Message)
async def delete_embedding_config(session: SessionDep, current_user: CurrentUser, config_id: uuid.UUID) -> Message:
    """Delete embedding configuration"""
    config = await session.get(EmbeddingConfigModel, config_id)
    if not config:
        raise NotFoundError("Configuration not found")

    if config.user_id != current_user.id:
        raise NotFoundError("Configuration not found")

    await session.delete(config)
    await session.commit()

    logger.info(f"Deleted embedding config {config_id} for user {current_user.id}")

    return Message(message="Configuration deleted successfully")


@router.post("/{config_id}/set-default", response_model=Message)
async def set_default_embedding_config(session: SessionDep, current_user: CurrentUser, config_id: uuid.UUID) -> Message:
    """Set default embedding configuration"""
    config = await session.get(EmbeddingConfigModel, config_id)
    if not config:
        raise NotFoundError("Configuration not found")

    if config.user_id != current_user.id:
        raise NotFoundError("Configuration not found")

    # Cancel other default configurations for the current user
    statement = select(EmbeddingConfigModel).where(
        EmbeddingConfigModel.user_id == current_user.id, EmbeddingConfigModel.is_default
    )
    existing_defaults = (await session.execute(statement)).scalars().all()
    for default_config in existing_defaults:
        default_config.is_default = False
        session.add(default_config)

    # Set new default configuration
    config.is_default = True
    session.add(config)
    await session.commit()

    return Message(message="Default configuration set successfully")


@router.get("/default/current", response_model=EmbeddingConfigPublic | None)
async def get_default_embedding_config(session: SessionDep, current_user: CurrentUser) -> Any:
    """Get default embedding configuration for the current user"""
    statement = select(EmbeddingConfigModel).where(
        EmbeddingConfigModel.user_id == current_user.id,
        EmbeddingConfigModel.is_default,
        EmbeddingConfigModel.is_active,
    )
    config = (await session.execute(statement)).scalars().first()

    return config
